pkg/storage/cacher.go

```diff
@@ -402,10 +403,20 @@ func filterFunction(key string, keyFunc func(runtime.Object) (string, error), fi
 			glog.Errorf("invalid object for filter: %v", obj)
 			return false
 		}
-		if !strings.HasPrefix(objKey, key) {
-			return false
+
+		segments := strings.Split(objKey, "/")
```

Comment:
```
if !strings.HasPrefix(objKey+"/", key) {
        return false
}
```

is also incorrect: if client watch "/pods/ns1/pod1", cache will NOT return events about "/pods/ns1/pod1", since objKey+"/" ="ns1/pod/", which is not a prefix of key "ns1/pod"